test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java

```diff
@@ -176,83 +179,177 @@ public TestCase nodes(int nodeCount)
+         * - v3x -> v41
+         * - v40 -> v41
+         * with CURRENT = v3x and minimalSupportedVersion = v30 the test case will run:
+         * - v30 -> v3x
+         */
+        public TestCase minimalApplicableVersion(Semver minimalSupportedVersion)
         {
-            return upgradesTo(from, CURRENT);
+            this.minimalApplicableVersion = minimalSupportedVersion;
+            addUpgrades();
+            return this;
         }
 
         /**
-         * performs all supported upgrade paths to the "to" target; example
-         * {@code upgradesTo(3.0, 4.0); // produces: 3.0 -> 4.0, 3.11 -> 4.0}
+         * @param minimalSupportedVersion the minimal version that this test case is applicable for
+         * @param maximalSupportedVersion the maximal version that this test case is applicable for
+         *
+         * This function sets both the lower and the upper bounds of test case's applicability
+         * the test case will run (see {@link #run}) all supported direct upgrades source -> target
+         * such that:
+         * - source >= minimalSupportedVersion
+         * - target <= maximalSupportedVersion
+         * - source == CURRENT or target == CURRENT
+         *
+         * example:
+         * with CURRENT = v50, minimalSupportedVersion = v30, and maximalSupportedVersion = v41
+         * the test case will not run any upgrades.
+         * with CURRENT = v41, minimalSupportedVersion = v30, and maximalSupportedVersion = v41
+         * the test case will run:
+         * - v30 -> v41
+         * - v3x -> v41
+         * - v40 -> v41
          */
-        public TestCase upgradesTo(Semver from, Semver to)
+        public TestCase applicableVersionsRange(Semver minimalSupportedVersion, Semver maximalSupportedVersion)
         {
-            List<TestVersions> upgrade = new ArrayList<>();
-            NavigableSet<Semver> vertices = sortedVertices(SUPPORTED_UPGRADE_PATHS);
-            for (Semver start : vertices.subSet(from, true, to, false))
-            {
-                // only include pairs that are allowed, and start or end on CURRENT
-                if (SUPPORTED_UPGRADE_PATHS.hasEdge(start, to) && contains(start, to, CURRENT))
-                    upgrade.add(new TestVersions(versions.getLatest(start), Collections.singletonList(versions.getLatest(to))));
-            }
-            logger.info("Adding upgrades of\n{}", upgrade.stream().map(TestVersions::toString).collect(Collectors.joining("\n")));
-            this.upgrade.addAll(upgrade);
+            this.minimalApplicableVersion = minimalSupportedVersion;
+            this.maximalApplicableVersion = maximalSupportedVersion;
+            addUpgrades();
             return this;
         }
 
         /**
-         * performs all supported upgrade paths from the "from" target; example
-         * {@code upgradesFrom(4.0, 4.2); // produces: 4.0 -> 4.1, 4.0 -> 4.2}
-         */
-        public TestCase upgradesFrom(Semver from, Semver to)
```

Comment:
likewise wrt javadoc

Reply:
the apidoc is correct. it is the `contains(..)` method that is wrong.

Reply:
no it isn't. the function also filters using `CURRENT`